File: longestSubstringWithoutRepeatingCharacters.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        start = 0
        maxl = 0
        d = {}
        for end in range(0, len(s)):
            curr = s[end]
            if curr in d:
                start = max(start, d[curr] + 1)
            d[curr] = end 
            maxl = max(maxl, end - start + 1)
        return maxl

File: longestSubstringWithoutRepeatingCharacters.go
package main

import "strconv"

func lengthOfLongestSubstring(s string) int {
	maxlen := 0
	start := 0
	subs := make(map[string]int)
	for end := range s {
		curr := strconv.Itoa(int(s[end]))
		_, ok := subs[curr]
		if ok {
			start = max(start, subs[curr]+1)
		}
		subs[curr] = end
		maxlen = max(maxlen, end-start+1)
	}
	return maxlen
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}
